pkg/runtime/grpc/oci.go

```diff
@@ -0,0 +1,274 @@
+	targets []target,
+) (runtime.CombinedGadgetResult, error) {
+	results := make(runtime.CombinedGadgetResult, len(targets))
+	var resultsLock sync.Mutex
+
+	wg := sync.WaitGroup{}
+	for _, t := range targets {
+		wg.Add(1)
+		go func(target target) {
+			gadgetCtx.Logger().Debugf("running oci gadget on node %q", target.node)
+			res, err := r.runOCIGadget(gadgetCtx, target, paramMap)
+			resultsLock.Lock()
+			results[target.node] = &runtime.GadgetResult{
+				Payload: res,
+				Error:   err,
+			}
+			resultsLock.Unlock()
+			wg.Done()
+		}(t)
+	}
+
+	wg.Wait()
+	return results, results.Err()
+}
+
+func (r *Runtime) runOCIGadget(gadgetCtx runtime.GadgetContext, target target, allParams map[string]string) ([]byte, error) {
+	// Notice that we cannot use gadgetCtx.Context() here, as that would - when cancelled by the user - also cancel the
+	// underlying gRPC connection. That would then lead to results not being received anymore (mostly for profile
+	// gadgets.)
+	connCtx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	timeout := time.Second * time.Duration(r.globalParams.Get(ParamConnectionTimeout).AsUint16())
+	dialCtx, cancelDial := context.WithTimeout(gadgetCtx.Context(), timeout)
+	defer cancelDial()
+
+	conn, err := r.dialContext(dialCtx, target, timeout)
+	if err != nil {
+		return nil, fmt.Errorf("dialing target on node %q: %w", target.node, err)
+	}
+	defer conn.Close()
+	client := api.NewOCIGadgetManagerClient(conn)
+
+	runRequest := &api.OCIGadgetRunRequest{
+		ImageName:   gadgetCtx.ImageName(),
+		ParamValues: allParams,
+		Args:        gadgetCtx.Args(),
+		LogLevel:    uint32(gadgetCtx.Logger().GetLevel()),
+		Timeout:     int64(gadgetCtx.Timeout()),
+		Version:     api.VersionGadgetRunProtocol,
+	}
+
+	runClient, err := client.RunOCIGadget(connCtx)
+	if err != nil && !errors.Is(err, context.Canceled) {
+		return nil, err
+	}
+
+	controlRequest := &api.OCIGadgetControlRequest{Event: &api.OCIGadgetControlRequest_OciRunRequest{OciRunRequest: runRequest}}
+	err = runClient.Send(controlRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	doneChan := make(chan error)
+
+	var result []byte
+	expectedSeq := uint32(1)
+
+	go func() {
+		dsMap := make(map[uint32]datasource.DataSource)
+		dsNameMap := make(map[string]uint32)
+		initialized := false
+		for {
+			ev, err := runClient.Recv()
+			if err != nil {
+				gadgetCtx.Logger().Debugf("%-20s | runClient returned with %v", target.node, err)
+				if !errors.Is(err, io.EOF) {
+					doneChan <- err
+					return
+				}
+				doneChan <- nil
+				return
+			}
+			switch ev.Type {
+			case api.EventTypeGadgetPayload:
+				if !initialized {
+					gadgetCtx.Logger().Warnf("%-20s | received payload without being initialized", target.node)
+					continue
+				}
+				if expectedSeq != ev.Seq {
+					gadgetCtx.Logger().Warnf("%-20s | expected seq %d, got %d, %d messages dropped", target.node, expectedSeq, ev.Seq, ev.Seq-expectedSeq)
+				}
+				expectedSeq = ev.Seq + 1
```

Comment:
On the sending side, the seq is set and the message forwarded to the output queue inside a lock. From there it's pushed synchronously to the TCP stack via gRPC. And since it's using TCP, we can expect to receive it in the same order or the connection will break.